### PR TITLE
[FLINK-12409][python] Adds from_elements in TableEnvironment

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request adds the API from_elements in TableEnvironment. It's a convenient API to create a table from a collection of elements.*
It works as follows:
1) Serializes the python data to a local file via PickleSerializer
2) Reads the local file and deserializes the python object and creates a DataStream from the deserialized elements in Java
3) Creates a Table from the created DataStream

## Brief change log

  - *Adds from_elements API in TableEnvironment*
  - *Adds PythonUtil to create a DataStream/DataSet from a file containing the serialized python objects*
  - *Adds PythonTableUtil to create a Table from a DataStream*
  - *Adds PythonTypeUtil to convert the deserialized objects according to the schema*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit test test_calc.test_from_elements*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by sunjincheng121** [comment]
The CI  throws `TypeError` for a python test case. I have been restarted it. And make sure all the test cases work well before the review. 
@flinkbot approve-until

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/typeutils/PythonTypeUtil.scala [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+import java.nio.charset.StandardCharsets
+import java.sql.{Date, Time, Timestamp}
+import java.util.function.BiConsumer
+
+import org.apache.flink.api.common.typeinfo.{BasicArrayTypeInfo, BasicTypeInfo, PrimitiveArrayTypeInfo, TypeInformation}
+import org.apache.flink.api.java.typeutils.{MapTypeInfo, ObjectArrayTypeInfo, RowTypeInfo}
+import org.apache.flink.table.api.Types
+import org.apache.flink.types.Row
+
+object PythonTypeUtil {
```
Comment (sunjincheng121) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/typeutils/PythonTypeUtil.scala]: Can we place this until in  `org.apache.flink.table.python`? 
BTW: due to this class only contains one public method `convertTo` and can not be used by JAVA users, so I suggest move the `convertTo` into `PythonTableUtil` and as a` private` method., What do you think?

**Comment by aljoscha** [comment]
I agree with @sunjincheng121, we must not put Python-relates code into unrelated packages.

**Comment by sunjincheng121** [comment]
Thanks for the reply @aljoscha!

As we(@aljoscha @dianfu and me) discussed offline, we got a consistent solution, we will put Python related Java code into
`flink-python` module, and we will add classifier named `java-binding` for the JAR. So, I appreciate if you can update the PR according to our decision. @dianfu  :)
Best, Jincheng

**Comment by dianfu** [comment]
Sounds good. I will rebase the PR.

File: flink-python/pyflink/table/table_environment.py [hunk 2]
```diff
@@ -379,13 +383,97 @@ def create(cls, table_config):
 
         return t_env
 
+    def from_elements(self, elements, schema=None, verify_schema=True):
+        """
+        Creates a table from a collection of elements.
+
+        :param elements: The elements to create table from.
+        :param schema: The schema of the table.
+        :param verify_schema: Whether to verify the elements against the schema.
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/table/table_environment.py]: If the user is confident that all the elements are consistent with the schema, there is no need to verify the schema.

File: flink-python/pyflink/table/table_environment.py [hunk 3]
```diff
@@ -379,13 +383,97 @@ def create(cls, table_config):
...
+                data_type, name="field value") if verify_schema else lambda _: True
+
+            def verify_obj(obj):
+                verify_func(obj)
+                return obj
+        else:
+            def verify_obj(obj):
+                return obj
+
+        # makes sure we distribute data evenly if it's smaller than self.batchSize
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/table/table_environment.py]: Good catch. Will remove this comment. Originally there is code corresponding to this comment, just forgot to remove this comment when removing that part of code.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/python/pickle/ArrayConstructor.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Creates arrays of objects. Returns a primitive type array such as int[] if
+ * the objects are ints, etc. Returns an ArrayList if it needs to
+ * contain arbitrary objects (such as lists).
+ *
+ * <p>Unpickle array.array generated by Python 2.6.
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/python/pickle/ArrayConstructor.java]: I see you point. +1 to remove the code handling Python <= 2.6.

**Comment by dianfu** [comment]
@sunjincheng121 Have addressed your comments and also moved all the Python related Java files to module flink-python.

**Comment by sunjincheng121** [comment]
@flinkbot attention @zentol

File: licenses-binary/LICENSE.py4j [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
```
[the remark this answers is not in the document]
Comment (dianfu) [reply licenses-binary/LICENSE.py4j]: Good catch. I think we should as currently we bundle the py4j-xxx-src.zip into the source distribution.

File: docs/ops/cli.md [hunk 6]
```diff
@@ -100,40 +100,40 @@ These examples about how to submit a job in CLI.
 
 -   Run Python Table program:
 
-        ./bin/flink run -py examples/python/table/batch/word_count.py -j <path/to/flink-table.jar>
+        ./bin/flink run -py examples/python/table/batch/word_count.py -C <path/to/flink-table.jar> -C <path/to/flink-python-*-java-binding.jar>
```
[the remark this answers is not in the document]
Comment (dianfu) [reply docs/ops/cli.md]: Make sense to me as flink-python-*-java-binding.jar is always needed for Python jobs.

**Comment by dianfu** [comment]
@sunjincheng121 Thanks a lot for your review. Have updated the PR accordingly.

**Comment by sunjincheng121** [comment]
Thanks for the quick updated! @dianfu 
LGTM. 
From my side, +1 to merged.

**Comment by sunjincheng121** [comment]
Thanks for the rebase, Merging...

File: NOTICE-binary [hunk 7]
```diff
@@ -5866,3 +5866,42 @@ This project bundles the following dependencies under the Apache Software Licens
...
+// Version 2.0, in this case for Apache Flink
+// ------------------------------------------------------------------
+
+Apache Flink
+Copyright 2006-2019 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+flink-python
```
Comment (zentol) [inline NOTICE-binary]: we already have a flink-python module in flink-dist; so we now have ambiguous licensing declarations. @sunjincheng121

Comment (sunjincheng121) [reply NOTICE-binary]: Thanks for point out this @zentol!
In our ( @aljoscha and me ) plan, we want to remove the `flink-libraries/flink-python` later, currently, we add the classifier “java-binding”  for the JAR.  for the licensing I agree with you, we really have ambiguous licensing declarations here.  :( 
We carefully thought about whether to change the module name, but the name flink-python is too apt, other names are not suitable, and the long-term goal we will remove the `flink-libraries/flink-python`. So we still want to use the name` flink-python`. So in order to eliminate ambiguity, can we add a little explanation in licensing? Or do you have better suggestions?

Comment (zentol) [reply NOTICE-binary]: well you could just call it flink-api-python or something, that's doesn't seem that inappropriate.

I don't believe we should be re-using artifactIds between releases if they are fundamentally different artifacts; it is for the same reason that we have not added another `flink-ml` module but named it `flink-ml-parent` instead.

Comment (sunjincheng121) [reply NOTICE-binary]: Currently, the names of modules in flink is `flink-language` patterns, such as `flink-java`, `flink-scala`. for the goal of `flink-python` will add  `user-defined-functions` support in the next release,  for UDFs we will add `data-service`, `state-service`...etc.  And it's not only the API level. These functions need to be integrated with the beam. The code inside flink will not be very much, so the current plan is to distinguish these codes with packages or folder(for python). So, If we should not re-using artifactId, how about using `flink-py`. @zentol @aljoscha  

BTW: I appreciate if we can discuss this module issue in the [design doc](https://docs.google.com/document/d/1ybYt-0xWRMa1Yf5VsuqGRtOfJBz4p74ZmDxZYg3j_h8/edit?usp=sharing):  https://docs.google.com/document/d/1ybYt-0xWRMa1Yf5VsuqGRtOfJBz4p74ZmDxZYg3j_h8/edit?usp=sharing

**Pull request closed without merging**